### ISSUE TITLE
BUG:  Unexpected change of behavior on DataFrame type float32 between pandas versions.

### ISSUE THREAD

**Issue by carlosan1708** [pandas-dev/pandas#46552]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
import json

def check_error(x, r):
    df = pd.DataFrame(data=x, dtype="float32")
    for i in df.index:
        for j in range(len(list(df.iloc[i]))):
            # Here is the difference, it"s converting str to float32 in pandas 1.3.5
            # On pandas 1.4.0 no longer happens, instead it's keeping the value intact.
            df.iloc[i][j] = r[j][int(df.iloc[i][j])]
    return df


x = [[0, 0.], [1., 0.], [0., 1.], [1., 1.]]
r = np.array([["10", "20"], ["50", "40"]])

result = check_error(x, r).to_json()
result_json = expected = json.loads(result)
## Pandas 1.3.5
if pd.__version__=="1.3.5":
    expected_json_pandas_1_3_5 = json.loads("""
    {"0":{"0":10.0,"1":20.0,"2":10.0,"3":20.0},"1":{"0":50.0,"1":50.0,"2":40.0,"3":40.0}}
    """)
    print(sorted(result_json.items()) == sorted(expected_json_pandas_1_3_5.items()))
else:
    expected_json_pandas_1_4_0 = json.loads("""
    {"0": {"0": 0.0, "1": 1.0, "2": 0.0, "3": 1.0}, "1": {"0": 0.0, "1": 0.0, "2": 1.0, "3": 1.0}}
    """)
    ## Pandas 1.4.0
    print(sorted(result_json.items()) == sorted(expected_json_pandas_1_4_0.items()))
```

### Issue Description

There is a change of behavior that is not mentioned in the documentation that could cause issues in existing libraries when assigning variables in a data frame with float32, I haven't checked if in other types of data frames the same could occur.

Pretty when using data frames with type float32 the assignation is not reacting in the same way between 1.3.5 pandas version and 1.4.0. In 1.3.5 str type is getting transformed into a float type, in 1.4.0 this is no longer occurring and instead, the assignation is not occurring anymore, but this is not throwing an error either which is causing as in the example two data frames to contain different information based on pandas version.

### Expected Behavior

Change should be called out in the documentation of 1.4.0 or throw an error instead to alert users about incorrect types during the assignation of variables. Overall it might be worthy to explore if this is not affecting other dataframes types. 

### Installed Versions

<details>

Pandas 1.3.5 

Pandas 1.4.0

Pandas 1.4.1

</details>

**Comment by phofl** [comment]
Please provide a minimal and reproducible example. You are using a lot of things, that should not be relevant to reproducing the bug. This makes it hard to understand.

See https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by carlosan1708** [comment]
@phofl Let me know if the example is good enough and if I need to update the code on top. 

Simplified version: 
```
import pandas as pd
import numpy as np
import json

# Assignation is the problem.
# In Pandas 1.3.5 matrix value gets substituted and in pandas 1.4.0 is not ocurring anymore.
df = pd.DataFrame(data=[[0]], dtype="float32")
r = np.array(["10"])
df.iloc[0][0] = r[int(df.iloc[0])]

comparable_result = df.to_json()
comparable_result_json = expected = json.loads(comparable_result)

#Comparison code.
## Pandas 1.3.5
if pd.__version__=="1.3.5":
    expected_json_pandas_1_3_5 = json.loads("""
    {"0":{"0":10.0}}
    """)
    print(sorted(comparable_result_json.items()) == sorted(expected_json_pandas_1_3_5.items()))
else: #Version 1.4.1 and 1.4.0 Tested.
    ## Result of the array is different.
    expected_json_pandas_1_4_0 = json.loads("""
    {"0":{"0":0.0}}
    """)
    ## Pandas 1.4.0
    print(sorted(comparable_result_json.items()) == sorted(expected_json_pandas_1_4_0.items()))

# Note: With array, behavior is not reproducible with the matrix is.

```

**Comment by simonjayhawkins** [comment]
I think the change in behavior is perhaps more clearly shown with

```
print(pd.__version__)
df = pd.DataFrame([[0, 0.0], [1.0, 0.0], [0.0, 1.0], [1.0, 1.0]], dtype="float32")
df.iloc[0][0] = "10"
print(df)
```
```
1.4.2
     0    1
0  0.0  0.0
1  1.0  0.0
2  0.0  1.0
3  1.0  1.0
```
```
1.3.5
      0    1
0  10.0  0.0
1   1.0  0.0
2   0.0  1.0
3   1.0  1.0
```

**Comment by jreback** [comment]
this should be showing a SettingWithCopyWarning with but maybe wasn't originally

**Comment by simonjayhawkins** [comment]
yep, no warning just silently does nothing now.

**Comment by simonjayhawkins** [comment]
@carlosan1708 Thanks for the report.

> In 1.3.5 str type is getting transformed into a float type, in 1.4.0 this is no longer occurring and instead, the assignation is not occurring anymore, but this is not throwing an error either which is causing as in the example two data frames to contain different information based on pandas version.

first bad commit: [fb9f205a43b683051ba61849e5a4cf6d99679171] BUG: Series.__setitem__ failing to cast numeric values (#45121)

> Change should be called out in the documentation of 1.4.0 or throw an error instead to alert users about incorrect types during the assignation of variables.

It doesn't look like this change was intentional. @jbrockmendel wdyt?

**Comment by jbrockmendel** [comment]
> In 1.3.5 str type is getting transformed into a float type, in 1.4.0 this is no longer occurring and instead

I think in other setitem-like ops we do not cast `"10"` to `float("10")`, so would expect `df.iloc[0, 0] = "10"` to the first column to object.  Having `df.iloc[0][0] = "10"` be a no-op makes sense ex the missing warning.

**Comment by simonjayhawkins** [comment]
yes, if the old behavior is considered a bug we do not necessarily need to restore it.

> Having `df.iloc[0][0] = "10"` be a no-op makes sense ex the missing warning.

hmm, what should be the behavior of `df.iloc[0][0] = 10`? because this still updates (and no warning).

**Comment by phofl** [comment]
Since 10 fits into the dtype, this can be done inplace, hence this works ( I think :))

**Comment by phofl** [comment]
@jbrockmendel Quick question:

```
na = np.array([[0, 0.0], [1.0, 0.0]])

df = pd.DataFrame(na)
x = df[0]
x[0] = 10
```

This changes na, because we are doing a Series.__setitem__ call which is a view on the DataFrame column. Is this expected?
```
[[10.  0.]
 [ 1.  0.]]
```

**Comment by simonjayhawkins** [comment]
> There is a change of behavior that is not mentioned in the documentation that could cause issues in existing libraries when assigning variables in a data frame with float32, I haven't checked if in other types of data frames the same could occur.

yes this occurs for float64 and int64 so assume applies to any numeric dtype.

so the title of this issue should perhaps be along the lines of "REGR: DataFrame with a numeric dtype no longer updates when assigning a numeric-like string value using chained indexing"

**Comment by jbrockmendel** [comment]
> This changes na, because we are doing a Series.setitem call which is a view on the DataFrame column. Is this expected?

Yes

**Comment by simonjayhawkins** [comment]
moving to 1.4.4

**Comment by simonjayhawkins** [comment]
> I think in other setitem-like ops we do not cast `"10"` to `float("10")`

continuing with the code sample from https://github.com/pandas-dev/pandas/issues/46552#issuecomment-1088584797

the actual change between 1.3.5 and 1.4 is

```python
print(pd.__version__)
df = pd.DataFrame([[0, 0.0], [1.0, 0.0], [0.0, 1.0], [1.0, 1.0]], dtype="float32")
ser = df.iloc[0]
ser[0] = "10"
print(ser)
print(type(ser[0]))
```

```
1.3.5
0    10.0
1     0.0
Name: 0, dtype: float32
<class 'numpy.float32'>
```

```
1.4.3
0     10
1    0.0
Name: 0, dtype: object
<class 'str'>
```

now, the release note for #45121 is about precision and upcasting, so I am not sure that this change of behavior was expected.

so ignoring the copy/view semantics for a moment, is this new behavior for the Series assignment correct?

**Comment by phofl** [comment]
Yes, imo this is correct

**Comment by simonjayhawkins** [comment]
> Yes, imo this is correct

and to be clear, this was a bug and the old behavior should not have first been deprecated?

**Comment by phofl** [comment]
Yes I think the old behavior was a bug. Setting „10“ into a column should behave the same as „a“ imo.
